File: R/mutual.R
mutual <- function(series, partitions=16, lag.max=20) {
	series <- (series-min(series))/(diff(range(series)))
	corr <- numeric(lag.max+1)
	for(i in 0:lag.max) {
		hist <- matrix(0, partitions, partitions)
		hist <- .C("mutual", series=as.double(series), length=as.integer(length(series)), lag=as.integer(i), partitions=as.integer(partitions), hist=as.double(hist), PACKAGE="tseriesChaos")[["hist"]]
		hist <- matrix(hist, partitions, partitions)/sum(hist)
		histx <- apply(hist, 1, sum)
		hist <- hist[hist!=0]
		histx<- histx[histx!=0]
		corr[i+1] <- sum(hist*log(hist)) - 2*sum(histx*log(histx))
	}
	names(corr) <- paste(0:lag.max)
	corr
}

File: src/mutual.c
#include "tseriesChaos.h"

#define output(i, j) out_hist[(i)*partitions + (j)]

void mutual(double *in_series, int *in_length, int *in_lag, 
int *in_partitions, double *out_hist) {
	int partitions, length, lag;
	int ix, iy, binx, biny, i, j;
	double *series;

	series = in_series;
	length = *in_length;
	lag =*in_lag;
	partitions=*in_partitions;

	for(i =0; i<partitions; i++) 
		for(j=0; j<partitions; j++)
			output(i, j) = 0.0;

	for(ix = 0; ix < (length-lag); ix++) {
		iy = ix + lag;
		binx = MIN((int)(series[ix]*partitions),partitions-1);
		biny = MIN((int)(series[iy]*partitions),partitions-1);
		output(binx, biny) ++;
	}
}
